### PR TITLE
Add uninstallpcackages

### PR DESCRIPTION
**Pull request by KazuCocoa**
Would like to add a new cap which uninstalled packages in initialising phase.
The use case is similar to `otherApps`, Appium installs specified packages by it.

### CODE REVIEW

**Comment by KazuCocoa** [comment]
Espresso driver can handle only one app. Thus, this feature is enough for UIA1/2

File: lib/android-helpers.js [hunk 1]
```diff
@@ -414,6 +414,19 @@ helpers.installOtherApks = async function (otherApps, adb, opts) {
...
 
+/**
+ * Uninstall an array of packages
+ * @param {[string]} appPackages An array of package names to uninstall
+ * @param {ADB} adb Instane of Appium ADB object
+ */
+helpers.uninstallOtherPackages = async function (appPackages, adb) {
+  logger.debug(`Uninstalling packages: ${appPackages}`);
+
+  for (const appPackage of appPackages) {
```
Comment (mykola-mokhnach) [inline lib/android-helpers.js]: I think this action could be paralleled. Consider using B.all

**Comment by KazuCocoa** [comment]
I'm adding to remove all 3rd party packages when a user set `*`. Will update this PR soon.

File: lib/android-helpers.js [hunk 2]
```diff
@@ -414,6 +414,19 @@ helpers.installOtherApks = async function (otherApps, adb, opts) {
   }));
 };
 
+/**
+ * Uninstall an array of packages
+ * @param {[string]} appPackages An array of package names to uninstall
+ * @param {ADB} adb Instane of Appium ADB object
```
Comment (imurchie) [inline lib/android-helpers.js]: Instance.

File: test/unit/driver-specs.js [hunk 3]
```diff
@@ -293,6 +293,40 @@ describe('driver', function () {
       await driver.initAUT();
       mocks.helpers.verify();
     });
+    it('should uninstall a package "uninstallOtherPackages" if set in capabiliteis', async function () {
```
Comment (imurchie) [inline test/unit/driver-specs.js]: `capabilities` (throughout).

File: lib/android-helpers.js [hunk 4]
```diff
@@ -414,6 +414,36 @@ helpers.installOtherApks = async function (otherApps, adb, opts) {
   }));
 };
 
+/**
+ * Uninstall an array of packages
+ * @param {[string]} appPackages An array of package names to uninstall. If this includes `'*'`, uninstall all of 3rd party apps
+ * @param {ADB} adb Instance of Appium ADB object
+ * @param {[string]} filterPackages An array of packages filter with when `*` is provided
```
Comment (mykola-mokhnach) [inline lib/android-helpers.js]: `{Array<string>}`

File: lib/android-helpers.js [hunk 5]
```diff
@@ -414,6 +414,36 @@ helpers.installOtherApks = async function (otherApps, adb, opts) {
...
+ */
+helpers.uninstallOtherPackages = async function (appPackages, adb, filterPackages = []) {
+  if (_.intersection(appPackages, ['*']).length === 1) {
```
Comment (mykola-mokhnach) [inline lib/android-helpers.js]: appPackages.includes('*')

File: lib/android-helpers.js (hunk 4 above)
Comment (mykola-mokhnach) [reply lib/android-helpers.js]: The current description is not very clear

File: lib/android-helpers.js [hunk 6]
```diff
@@ -414,6 +414,36 @@ helpers.installOtherApks = async function (otherApps, adb, opts) {
...
+    appPackages = await this._get3rdPartyPackages(adb, filterPackages);
+  }
+
+  logger.debug(`Uninstalling packages: ${appPackages}`);
+  await B.all(appPackages.map((appPackage) => {
+    return adb.uninstallApk(appPackage);
+  }));
+};
+
+helpers._get3rdPartyPackages = async function (adb, filterPackages = []) {
```
Comment (mykola-mokhnach) [inline lib/android-helpers.js]: `getThirdPartyPackages`. There is no need to make this method private

File: lib/android-helpers.js [hunk 7]
```diff
@@ -414,6 +414,36 @@ helpers.installOtherApks = async function (otherApps, adb, opts) {
...
+  try {
+    const packages = await adb.shell(['pm', 'list', 'packages', '-3']);
+    const appPackages = packages.trim().replace(/package:/g, '').split('\n');
```
Comment (mykola-mokhnach) [inline lib/android-helpers.js]: split(os.EOL)

Comment (mykola-mokhnach) [reply lib/android-helpers.js]: or add `.map((x) -> x.trim())`

File: lib/android-helpers.js [hunk 8]
```diff
@@ -414,6 +414,36 @@ helpers.installOtherApks = async function (otherApps, adb, opts) {
...
 
+/**
+ * Uninstall an array of packages
+ * @param {[string]} appPackages An array of package names to uninstall. If this includes `'*'`, uninstall all of 3rd party apps
+ * @param {ADB} adb Instance of Appium ADB object
+ * @param {[string]} filterPackages An array of packages filter with when `*` is provided
+ */
+helpers.uninstallOtherPackages = async function (appPackages, adb, filterPackages = []) {
+  if (_.intersection(appPackages, ['*']).length === 1) {
+    logger.debug('Uninstall 3rd party packages');
```
Comment (mykola-mokhnach) [inline lib/android-helpers.js]: third-party

File: lib/android-helpers.js [hunk 9]
```diff
@@ -414,6 +414,36 @@ helpers.installOtherApks = async function (otherApps, adb, opts) {
...
+    appPackages = await this._get3rdPartyPackages(adb, filterPackages);
+  }
+
+  logger.debug(`Uninstalling packages: ${appPackages}`);
+  await B.all(appPackages.map((appPackage) => {
```
Comment (mykola-mokhnach) [inline lib/android-helpers.js]: `B.all(appPackages.map((appPackage) => adb.uninstallApk(appPackage))`

File: lib/driver.js [hunk 10]
```diff
@@ -322,6 +322,18 @@ class AndroidDriver extends BaseDriver {
...
 
+    // Uninstall any uninstallOtherPackages which were specified in caps
+    if (this.opts.uninstallOtherPackages) {
+      let uninstallOtherPackages;
+      try {
+        uninstallOtherPackages = helpers.parseArray(this.opts.uninstallOtherPackages);
+      } catch (e) {
+        log.errorAndThrow(`Could not parse "uninstallOtherPackages" capability: ${e.message}`);
+      }
+      // Only SETTINGS_HELPER_PKG_ID package is used by UIA1
```
Comment (mykola-mokhnach) [inline lib/driver.js]: how about the bootstrap jar?

Comment (KazuCocoa) [reply lib/driver.js]: I thought so, but I was able to get only the setting package

File: lib/driver.js [hunk 11]
```diff
@@ -322,6 +322,18 @@ class AndroidDriver extends BaseDriver {
     Object.assign(this.opts, launchInfo);
     Object.assign(this.caps, launchInfo);
 
+    // Uninstall any uninstallOtherPackages which were specified in caps
+    if (this.opts.uninstallOtherPackages) {
+      let uninstallOtherPackages;
+      try {
+        uninstallOtherPackages = helpers.parseArray(this.opts.uninstallOtherPackages);
```
Comment (imurchie) [inline lib/driver.js]: I'd prefer this stuff be in the [validateDesiredCaps](https://github.com/appium/appium-android-driver/blob/master/lib/android-helpers.js#L702-L718) function.

File: lib/android-helpers.js [hunk 12]
```diff
@@ -733,6 +733,13 @@ helpers.parseArray = function (cap) {
   throw new Error(`must provide a string or JSON Array; received ${cap}`);
 };
 
+/**
+ * Validate desired capabilities. Returns true if capability is valid
+ *
+ * @param {any} cap A desired capability
```
Comment (mykola-mokhnach) [inline lib/android-helpers.js]: `{*}`

File: lib/android-helpers.js [hunk 13]
```diff
@@ -733,6 +733,13 @@ helpers.parseArray = function (cap) {
...
+ * @return {bool} Returns true if the capability is valid
```
Comment (mykola-mokhnach) [inline lib/android-helpers.js]: `{boolean}`

File: lib/android-helpers.js [hunk 14]
```diff
@@ -414,6 +414,40 @@ helpers.installOtherApks = async function (otherApps, adb, opts) {
   }));
 };
 
+/**
+ * Uninstall an array of packages
+ * @param {Array<string>} appPackages An array of package names to uninstall. If this includes `'*'`, uninstall all of 3rd party apps
+ * @param {ADB} adb Instance of Appium ADB object
+ * @param {Array<string>} filterPackages An array of packages does not uninstall when `*` is provided as `appPackages`
+ */
+helpers.uninstallOtherPackages = async function (appPackages, adb, filterPackages = []) {
```
Comment (mykola-mokhnach) [inline lib/android-helpers.js]: usually adb is the first parameter to all these methods

**Comment by KazuCocoa** [comment]
Could you publish this as a new version number?

**Pull request merged**